drivers/serial/uart_cdns.c

```diff
@@ -0,0 +1,305 @@
+}
+
+/** Check for IRQ updates */
+static int uart_cdns_update_irq(const struct device *dev)
+{
+	return 1;
+}
+
+/** Set the callback function pointer for IRQ. */
+void uart_cdns_set_irq_callback(const struct device *dev, uart_irq_callback_user_data_t cb,
+				   void *cb_data)
+{
+	struct uart_cdns_data *data;
+
+	data = DEV_DATA(dev);
+	data->callback = cb;
+	data->cb_data = cb_data;
+}
+
+static void uart_cdns_irq_handler(const struct device *dev)
+{
+	struct uart_cdns_regs *uart_regs = DEV_UART(dev);
+	uint32_t key = irq_lock();
+	uint32_t isr_status;
+	struct uart_cdns_data *data = DEV_DATA(dev);
+
+	if (data->callback) {
+		data->callback(dev, data->cb_data);
+	}
+
+	/* clear events */
+	/* need to make local copy of the status */
+	isr_status = uart_regs->channel_intr_status;
+
+	irq_unlock(key);
+}
+
+#endif
+
+static const struct uart_driver_api uart_cdns_driver_api = {
+	.poll_in = uart_cdns_poll_in,
+	.poll_out = uart_cdns_poll_out,
+	.err_check = NULL,
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+	.fifo_fill = uart_cdns_fill_fifo,
+	.fifo_read = uart_cdns_read_fifo,
+	.irq_tx_enable = uart_cdns_enable_tx_irq,
+	.irq_tx_disable = uart_cdns_disable_tx_irq,
+	.irq_tx_ready = uart_cdns_irq_tx_ready,
+	.irq_rx_enable = uart_cdns_enable_rx_irq,
+	.irq_rx_disable = uart_cdns_disable_rx_irq,
+	.irq_rx_ready = uart_cdns_irq_rx_ready,
+	.irq_err_enable = uart_cdns_enable_irq_err,
+	.irq_err_disable = uart_cdns_disable_irq_err,
+	.irq_is_pending = uart_cdns_is_irq_pending,
+	.irq_update = uart_cdns_update_irq,
+	.irq_callback_set = uart_cdns_set_irq_callback
+#endif
+};
+
+/** Initialize the UART */
+static int uart_cdns_init(const struct device *dev)
+{
+	struct uart_cdns_regs *uart_regs = DEV_UART(dev);
+	const struct uart_cdns_device_config *const dev_cfg = DEV_CFG(dev);
+
+	if (!dev_cfg->do_not_init) {
+		/* Reset RX and TX path */
+		uart_regs->ctrl = (CTRL_RXRES_MASK | CTRL_TXRES_MASK);
+
+		/* Disable TX and RX channels */
+		uart_regs->ctrl = (CTRL_STPBRK_MASK | CTRL_TXDIS_MASK | CTRL_RXDIS_MASK);
+
+		/* Configure Baud rate */
+		uart_cdns_set_baudrate(uart_regs, dev_cfg, dev_cfg->baud_rate);
+
+		/* Configure the mode */
+		uart_regs->mode = (SET_VAL32(MODE_WSIZE, 1) | SET_VAL32(MODE_UCLKEN, 1) |
+				   SET_VAL32(MODE_PAR, dev_cfg->parity));
+
+		/* Disable all interrupts */
+		uart_regs->intr_disable = 0xFFFFFFFF;
+
+		/* Enable TX and RX Channels */
+		uart_regs->ctrl = (CTRL_TXEN_MASK | CTRL_RXEN_MASK | CTRL_STPBRK_MASK);
+
+		if (dev_cfg->cfg_func) {
+			/* Setup IRQ handler */
+			dev_cfg->cfg_func();
+		}
+	}
+	return 0;
+}
+
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+
+#define UART_IRQ_PRIORITY(n)									   \
+	COND_CODE_1(DT_IRQ_HAS_CELL(DT_DRV_INST(n), priority), (DT_IRQ(DT_DRV_INST(n), priority)), \
+		    (0))
+
+#define UART_CDNS_IRQ_CFG_FUNC(n)								   \
+	static void uart_cdns_irq_cfg_func_##n(void)						   \
+	{											   \
+		IRQ_CONNECT(DT_INST_IRQN(n), UART_IRQ_PRIORITY(n), uart_cdns_irq_handler,	   \
```

Comment:
I would expect that priority should always be there. `DT_INST_IRQ` should be all you need
`IRQ_CONNECT(DT_INST_IRQN(n), DT_INST_IRQ(n, priority), uart_cdns_irq_handler,`

Reply:
Updated. Thanks.